File: 116-populating-next-right-pointers-in-each-node/116-populating-next-right-pointers-in-each-node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (!root)
            return root;

        std::queue<Node *> myQ;
    
        myQ.push(root);
        root->next = NULL;

        while (myQ.empty() != true)
        {
            int qSize = myQ.size();
            Node *prev = NULL;

            for(int i = 0; i < qSize; i++) {
                Node *cur = myQ.front();
                myQ.pop();

                if(cur->left)
                    myQ.push(cur->left);
                
                if(cur->right)
                    myQ.push(cur->right);
                if( prev )
                    prev->next = cur;
                prev = cur;
            }
        }

        return root;
 
    }
};

File: 116-populating-next-right-pointers-in-each-node/116-populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution(object):
    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        if root is None:
            return None
        
        q = collections.deque([root])
        while len(q) != 0:
            numNode = len(q)
            prev = None
            for _ in range(numNode):
                node = q.popleft()
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
                if prev is not None:
                    prev.next = node
                prev = node
        
        return root
        